src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -154,22 +153,25 @@ internal void Update<T>(T value, ReadOnlySpan<KeyValuePair<string, object>> tags
             }
         }
 
-        internal List<IMetric> Collect(bool isDelta)
+        internal List<IMetric> Collect(bool isDelta, DateTimeOffset dt)
         {
             var collectedMetrics = new List<IMetric>();
 
-            var dt = DateTimeOffset.UtcNow;
-
-            foreach (var keys in this.keyValue2MetricAggs)
+            // Lock to prevent new time series from being added
+            // until collect is done.
+            lock (this.lockKeyValue2MetricAggs)
             {
-                foreach (var values in keys.Value)
+                foreach (var keys in this.keyValue2MetricAggs)
                 {
-                    foreach (var metric in values.Value)
+                    foreach (var values in keys.Value)
                     {
-                        var m = metric.Collect(dt, isDelta);
-                        if (m != null)
+                        foreach (var metric in values.Value)
                         {
-                            collectedMetrics.Add(m);
+                            var m = metric.Collect(dt, isDelta);
```

Comment:
You may want to refactor to keep the metric.Collect() outside of the lock().

Not a blocker at this time.